src/PackageUtilities.js

```diff
@@ -356,40 +345,59 @@ export default class PackageUtilities {
   }
 
   /**
-   * Create a symlink to a dependency's binary in the node_modules/.bin folder
-   * @param {String} src
-   * @param {String} dest
-   * @param {String} name
-   * @param {String|Object} bin
+   * Symlink bins of srcPackage to node_modules/.bin in destPackage
+   * @param {Object|string} srcPackageRef
+   * @param {Object|string} destPackageRef
    * @param {Function} callback
    */
-  static createBinaryLink(src, dest, name, bin, callback) {
-    const safeName = name[0] === "@"
-      ? name.substring(name.indexOf("/") + 1)
-      : name;
-    const destBinFolder = path.join(dest, ".bin");
-
-    // The `bin` in a package.json may be either a string or an object.
-    // Normalize to an object.
-    const bins = typeof bin === "string"
-      ? { [safeName]: bin }
-      : bin;
-
-    const srcBinFiles = [];
-    const destBinFiles = [];
-    Object.keys(bins).forEach((binName) => {
-      srcBinFiles.push(path.join(src, bins[binName]));
-      destBinFiles.push(path.join(destBinFolder, binName));
-    });
+  static createBinaryLink(srcPackageRef, destPackageRef, callback) {
+    const srcPackage = resolvePackageRef(srcPackageRef);
+    const destPackage = resolvePackageRef(destPackageRef);
+
+    const bin = getPackageBin(srcPackageRef);
+
+    const actions = entries(bin)
+      .map(([name, file]) => ({
+        src: path.join(srcPackage.location, file),
+        dst: path.join(destPackage.binLocation, name)
+      }))
+      .reduce((acc, {src, dst}) => {
+        const link = (cb) => FileSystemUtilities.symlink(src, dst, "exec", cb);
+        const chmod = (cb) => FileSystemUtilities.chmod(src, "755", cb);
+        const actions = FileSystemUtilities.existsSync(src) ? [link, chmod] : [];
```

Comment:
I am an offender of using compiled output as `bin`. As it stands now I'd be deadlocked (without changes) in some of my lerna repositories when we decide to throw for non/not-yet existing bin targets.

I wonder if we could automate the proposed fix by placing the "proxy" files in the target `node_modules/.bin/` directories instead of symlinking? 

What do you think?

Reply:
I'm okay with skipping rather than erroring, for now. The proxy idea makes me nervous because we're diverging from the original intent of the package.json. It's a signal, to me, that we're heading in the wrong direction.